include/lock/optimistic_lock.hpp

```diff
@@ -0,0 +1,287 @@
+
+      std::this_thread::sleep_for(kShortSleep);
+    }
+  }
+
+  /**
+   * @brief Check whether the version after reading is equal to the version before reading
+   *
+   * @return true
+   * @return false
+   */
+  auto
+  CheckVersion(const uint64_t expected) const  //
+      -> bool
+  {
+    const auto desired = lock_.load(std::memory_order_relaxed) & kLockMask;
+    return expected == desired;
+  }
+
+  /**
+   * @brief Get an exclusive lock.
+   *
+   * NOTE: this function does not give up acquiring locks and continues spinlock.
+   */
+  void
+  LockX()
+  {
+    while (true) {
+      auto expected = lock_.load(std::memory_order_relaxed) & kLockMask;
+      auto desired = expected + kXLock;
+      for (size_t i = 0; i < kRetryNum; ++i) {
+        const auto cas_success = lock_.compare_exchange_weak(
+            expected, desired, std::memory_order_acquire, std::memory_order_relaxed);
+        if (cas_success) return;
+
+        expected &= kLockMask;
+        desired = expected + kXLock;
+        SPINLOCK_HINT
+      }
+
+      std::this_thread::sleep_for(kShortSleep);
+    }
+  }
+
+  /**
+   * @brief Downgrade an X lock to an SIX lock.
+   *
+   * NOTE: if a thread that does not have an exclusive lock calls this function, it will
+   * corrupt an internal lock status.
+   */
+  void
+  DowngradeToSIX()
+  {
+    lock_.fetch_add(kXLock + kSIXLock, std::memory_order_release);
+  }
+
+  /**
+   * @brief Release an exclusive lock.
+   *
+   */
+  void
+  UnlockX()
+  {
+    lock_.fetch_add(kXLock, std::memory_order_release);
+  }
+
+  /**
+   * @brief Get a shared lock with an intent-exclusive lock.
+   *
+   * NOTE: this function does not give up acquiring locks and continues spinlock.
+   */
+  void
+  LockSIX()
+  {
+    while (true) {
+      auto expected = lock_.load(std::memory_order_relaxed) & kXLockMask;
+      auto desired = expected + kSIXLock;
+      for (size_t i = 0; i < kRetryNum; ++i) {
+        const auto cas_success = lock_.compare_exchange_weak(
+            expected, desired, std::memory_order_acquire, std::memory_order_relaxed);
+        if (cas_success) return;
+
+        expected &= kXLockMask;
+        desired = expected + kSIXLock;
+        SPINLOCK_HINT
+      }
+
+      std::this_thread::sleep_for(kShortSleep);
+    }
+  }
+
+  /**
+   * @brief Upgrade an SIX lock to an X lock.
+   *
+   * NOTE: if a thread that does not have a shared lock with an intent-exclusive lock
+   * calls this function, it will corrupt an internal lock status.
+   */
+  void
+  UpgradeToX()
+  {
+    while (true) {
+      auto expected = (lock_ & kLockMask) + kSIXLock;
+      for (size_t i = 0; i < kRetryNum; ++i) {
+        const auto cas_success = lock_.compare_exchange_weak(
+            expected, kXLock, std::memory_order_acquire, std::memory_order_relaxed);
+        if (cas_success) return;
+
+        expected = (lock_ & kLockMask) + kSIXLock;
+        SPINLOCK_HINT
+      }
+
+      std::this_thread::sleep_for(kShortSleep);
+    }
```

Comment:
`desired`を別途設定してあげないとバージョン番号がリセットされる気がします．なお，`kSBitsMask`は共有ロックカウンタのみゼロにするマスクの想定なのと，SIXロックが取れた時点でフェンスが挟んであるのでここのCASはフェンスなしでいけます（SIXロックにより書込みなしが保証されるので）．

```suggestion
    while (true) {
      auto expected = lock_.load(std::memory_order_relaxed) & kSBitsMask;
      const auto desired = expected + kSIXLock;
      for (size_t i = 0; i < kRetryNum; ++i) {
        const auto cas_success = lock_.compare_exchange_weak(
            expected, desired, std::memory_order_relaxed);
        if (cas_success) return;

        SPINLOCK_HINT
      }

      std::this_thread::sleep_for(kShortSleep);
    }
```

Reply:
CASが失敗したときに以下の処理が必要な気がするのですが解釈合っているでしょうか？
```
        expected &= kSBitsMask;
        desired = expected + kSIXLock;
```

Reply:
たぶんなしでいけます．SIXロックを保持している以上，他のスレッドがXロックを取ることがないので，バージョン番号が増えることもありません．なので結局ここは共有ロックカウンタがゼロになるのを待ってるだけなので，最初に設定した`expected`と`desired`が使い回せる気がします．

Reply:
見直して気づきましたが，なので221-222行目はwhileループの外でも行ける気がします．

Reply:
変なことを言っていたら申し訳ないのですが，
CASが失敗したとき(共有ロックが取られている時)に，`expected`の値が置き換わって，ここでCASが成功すると`desired`に変更されるので共有ロックを無理やり0にするという状態にならないでしょうか？

`desired`は使いまわせそうです！

Reply:
ボケてました．CAS失敗時に`expected`は現在の値に書き換わるので，マスクの再適用は必要ですね．

Reply:
分かりました！ありがとうございます！